static/app/components/timezoneProvider.tsx

```diff
@@ -0,0 +1,64 @@
+import {createContext, useContext, useEffect, useMemo, useState} from 'react';
+
+import {useUser} from 'sentry/utils/useUser';
+
+interface TimezoneProviderValue {
+  /**
+   * A function that may be used to override the configured timezone. The
+   * timezone may be restored to the default (user configured timezone) by
+   * calling this function with null
+   */
+  setOverridenTimezone: (timezone: string | null) => void;
+  /**
+   * The current configured timezone stored in the provider
+   */
+  timezone: string;
+}
+
+interface TimezoneProviderProps {
+  children: NonNullable<React.ReactNode>;
+}
+
+const Provider = createContext<TimezoneProviderValue>({
+  timezone: 'Etc/UTC',
+  setOverridenTimezone: () => {},
+});
+
+const browserTimezone = Intl.DateTimeFormat().resolvedOptions().timeZone;
+
+/**
+ * Provides a timezone value to components that call useTimezone
+ */
+export function TimezoneProvider({children}: TimezoneProviderProps) {
+  const user = useUser();
+  const [overridenTimezone, setOverridenTimezone] = useState<string | null>(null);
+
+  const userTimezone = user?.options.timezone;
+
+  const timezone = overridenTimezone ?? userTimezone ?? browserTimezone;
+
+  const value = useMemo<TimezoneProviderValue>(
+    () => ({timezone, setOverridenTimezone}),
+    [timezone]
+  );
+
+  return <Provider value={value}>{children}</Provider>;
+}
+
+/**
+ * Get the currently configured timezone from the nearest TimezoneProvider.
+ */
+export function useTimezone() {
+  return useContext(Provider).timezone;
+}
+
+/**
+ * This hook may be used to override the configured timezone in the nearest
+ * TimezoneProvider. The timezone will be override will be removed when the
+ * component rendering this hook is unmounted, or the hook is provided `null`.
```

Comment:
```suggestion
 * This hook may be used to override the configured timezone in the nearest
 * TimezoneProvider. The override can be reverted by calling `useOverrideTimezone(null)`
```

Reply:
hmm, that feels unexpected to me. the way I reason about hooks at least is that they are active while they are mounted, unmounting a hook should not typically leave any side effects.

I think if we want that we could have additional options like

```tsx
useOverrideTimezone(timezone, {persist: true})
```

how does that feel?

Reply:
did you push `UserTimezoneProvider`?

Reply:
lol sorry I didn't. Let me actually re-request review in a little bit from you if that's OK